library/serialize.c

```diff
@@ -0,0 +1,325 @@
+            return( MBEDTLS_ERR_SERIALIZE_SEND );
+    do {
+        result = write( serialize_write_fd, buffer, length );
+        /* This channel should never reach EOF under the current simplistic
+           design, so error out on EOF. */
+        if( result <= 0 )
+        {
+            return( MBEDTLS_ERR_SERIALIZE_SEND );
+        }
+        length -= result;
+        buffer += result;
+    } while( length > 0 );
+    return( 0 );
+}
+
+static int mbedtls_serialize_read( uint8_t *buffer, size_t length )
+{
+    ssize_t result;
+    do {
+        result = read( serialize_read_fd, buffer, length );
+
+        /* This channel should never reach EOF under the current simplistic
+           design, so error out on EOF. */
+        if( result <= 0 )
+        {
+            return( MBEDTLS_ERR_SERIALIZE_RECEIVE );
+        }
+        length -= result;
+        buffer += result;
+    } while( length > 0 );
+    return( 0 );
+}
+
+#endif /* MBEDTLS_SERIALIZE_FORK_FRONTEND_C */
+
+int mbedtls_serialize_push_buffer( const void *buffer, size_t length )
+{
+    int ret;
+    uint8_t header[4];
+    if( length > MBEDTLS_SERIALIZE_MAX_STRING_LENGTH )
+        return( MBEDTLS_ERR_SERIALIZE_BAD_INPUT );
+
+    header[0] = MBEDTLS_SERIALIZE_TYPE_PUSH;
+    header[1] = ( length >> 16 ) & 0xff;
+    header[2] = ( length >> 8 ) & 0xff;
+    header[3] = length & 0xff;
+
+    if( ( ret = mbedtls_serialize_write( header, sizeof( header ) ) ) != 0 )
+        return( ret );
+    if( ( ret = mbedtls_serialize_write( buffer, length ) ) != 0 )
+        return( ret );
+    return( 0 );
+}
+
+int mbedtls_serialize_push_int16( uint16_t value )
+{
+    uint8_t buffer[2];
+    buffer[0] = ( value >> 8 ) & 0xff;
+    buffer[1] = value & 0xff;
+    return( mbedtls_serialize_push_buffer( buffer, sizeof( buffer ) ) );
+}
+
+int mbedtls_serialize_push_int32( uint32_t value )
+{
+    uint8_t buffer[4];
+    buffer[0] = ( value >> 24 ) & 0xff;
+    buffer[1] = ( value >> 16 ) & 0xff;
+    buffer[2] = ( value >> 8 ) & 0xff;
+    buffer[3] = value & 0xff;
+    return( mbedtls_serialize_push_buffer( buffer, sizeof( buffer ) ) );
+}
+
+int mbedtls_serialize_execute( uint32_t command )
+{
+    int ret;
+    uint8_t buffer[4];
+    buffer[0] = MBEDTLS_SERIALIZE_TYPE_EXECUTE;
+    buffer[1] = ( command >> 16 ) & 0xff;
+    buffer[2] = ( command >> 8 ) & 0xff;
+    buffer[3] = command & 0xff;
+    if( ( ret = mbedtls_serialize_write( buffer, sizeof( buffer ) ) ) != 0 )
+        return( ret );
+    if( ( ret = mbedtls_serialize_read( buffer, sizeof( buffer ) ) ) != 0 )
+        return( ret );
+    if( ! ( buffer[0] == MBEDTLS_SERIALIZE_TYPE_RESULT &&
+            buffer[1] == 0 && buffer[2] == 0 && buffer[3] == 4 ) )
+        return( MBEDTLS_ERR_SERIALIZE_BAD_OUTPUT );
+    if( ( ret = mbedtls_serialize_read( buffer, sizeof( buffer ) ) ) != 0 )
+        return( ret );
+    return( buffer[0] << 24 | buffer[1] << 16 | buffer[2] << 8 | buffer[3] );
```

Comment:
Please explicitly cast `buffer[0]` as `uint32_t` to avoid undefined behavior due to overly large shift (the operand `buffer[0]` is promoted to `int` before `<< 24` is applied).